src/target/riscv/riscv.c

```diff
@@ -3017,9 +3017,6 @@ static int riscv_mmu(struct target *target, int *enabled)
 {
 	*enabled = 0;
 
-	if (!riscv_virt2phys_mode_is_sw(target))
-		return ERROR_OK;
-
```

Comment:
Access memory would fail while virt2phys_mode is hw.
```suggestion
	if (riscv_virt2phys_mode_is_hw(target))
		return ERROR_OK;
```

Reply:
With this check `virt2phys_mode` would return incorrect value: virtual address instead of physical

Reply:
Access memory would fail without this check `virt2phys_mode` while virt2phys_mode is hw. It is correct to return a virtual address while virt2phys_mode is hw. Address translation should be done by hardware when set `dcsr.mprven`, I have tested it. When virt2phys_mode is hw, the virt2phys command returns the physical address, at least the current modification is incorrect, and both the virt2phys command and memory access will report errors.

Reply:
Sorry, I've mistaken. I meant that `virt2phys` would return incorrect value

Reply:
I think I understand you now. You are right, here we have double translation. But I think it has to be fixed in `riscv_rw_memory` function. Something like:
```c
if (!riscv_virt2phys_mode_is_hw(target))
	return r->access_memory(target, args);
```
before translation.

Sorry for the mess. This change was a part of #1241. Looks like I extracted this part incorrectly

Reply:
When virt2phys_mode is hw, the virt2phys command still report errors. Translate page table accesses failed due to setting `dcsr.mprven`. I think maybe it could be fixed in `riscv_virt2phys` function like this:
``` 
        RISCV_INFO(r);
        int mode = r->virt2phys_mode;
        if (riscv_virt2phys_mode_is_hw(target))
                r->virt2phys_mode = RISCV_VIRT2PHYS_MODE_SW;

        int result = riscv_address_translate(target,
                        satp_info, get_field(satp_value, RISCV_SATP_PPN(xlen)),
                        NULL, 0,
                        virtual, physical);

        r->virt2phys_mode = mode;
        return result;
```

Reply:
IIUC, riscv_address_translate() function should always perform physical accesses - it should work exclusively with physical addresses. @lz-bro, is this the issue that you are trying to point out?

If that's correct understanding, then we should probably create function `r->access_memory_phys` that will always perform physical accesses, regardless of `r->virt2phys_mode`. It looks to me as a cleaner solution than temporarily changing the `virt2phys_mode` flag.

Reply:
@JanMatCodasip Yes, that's exactly what I meant.

Reply:
@lz-bro, great catch!

> If that's correct understanding, then we should probably create function r->access_memory_phys that will always perform physical accesses, regardless of r->virt2phys_mode. It looks to me as a cleaner solution than temporarily changing the virt2phys_mode flag.

@JanMatCodasip, this will not be enough. In case of 2-stage translation, the first PTE is addressed using a virtual address. I'd suggest bailing out with a message that this is not supported for now (AFAIU, `mstatus.MPP` will need to be modified/restored for proper support).

> ... create function r->access_memory_phys that will always perform physical accesses.

AFAIU, this is exactly the purpose of #1241.

Reply:
> AFAIU, this is exactly the purpose of https://github.com/riscv-collab/riscv-openocd/pull/1241

You are right, yeah